interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,373 @@
+// returns it and does not set maxID.
+func (pdb *PromptDB) loadMaxID() error {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+
+	target := pdb.maxIDPath
+	f, err := os.Open(target)
+	if os.IsNotExist(err) {
+		pdb.maxID = 0
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("cannot read maximum prompt ID: %w", err)
+	}
+	defer f.Close()
+
+	idBuf := [16]byte{}
+	_, err = io.ReadFull(f, idBuf[:])
+	if err != nil {
+		return fmt.Errorf("cannot read maximum prompt ID: %w", err)
+	}
+
+	maxID, err := strconv.ParseUint(string(idBuf[:]), 16, 64)
+	if err != nil {
+		return fmt.Errorf("cannot parse maximum prompt ID: %w", err)
+	}
+
+	pdb.maxID = maxID
+	return nil
+}
+
+// nextID advances the internal monotonically-increasing maxID integer.
+//
+// The caller must ensure that the prompt DB mutex is held.
+func (pdb *PromptDB) nextID() string {
+	pdb.maxID++
+	idStr := pdb.maxIDString()
+	osutil.AtomicWriteFile(pdb.maxIDPath, []byte(idStr), 0600, 0)
+	return idStr
+}
+
+// maxIDString returns a 16-character string corresponding to the current maxID.
+// The ID string is the max ID in hexadecimal, padded by leading zeroes.
+func (pdb *PromptDB) maxIDString() string {
+	return fmt.Sprintf("%016X", pdb.maxID)
+}
+
+// AddOrMerge checks if the given prompt contents are identical to an existing
+// prompt and, if so, merges with it by adding the given listenerReq to it.
+// Otherwise, adds a new prompt with the given contents to the prompt DB.
+//
+// If the prompt was merged with an identical existing prompt, returns the
+// existing prompt and true, indicating it was merged. If a new prompt was
+// added, returns the new prompt and false, indicating the prompt was not
+// merged.
+//
+// The caller must ensure that the given permissions are in the order in which
+// they appear in the available permissions list for the given interface.
+func (pdb *PromptDB) AddOrMerge(user uint32, snap string, iface string, path string, permissions []string, listenerReq *listener.Request) (*Prompt, bool) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, ok := pdb.perUser[user]
+	if !ok {
+		pdb.perUser[user] = &userPromptDB{
+			ByID: make(map[string]*Prompt),
+		}
+		userEntry = pdb.perUser[user]
+	}
+
+	availablePermissions, _ := prompting.AvailablePermissions(iface)
+	// Error should be impossible, since caller has already validated that iface
+	// is valid, and tests check that all valid interfaces have valid available
+	// permissions returned by AvailablePermissions.
+
+	constraints := &promptConstraints{
+		Path:                 path,
+		Permissions:          permissions,
+		AvailablePermissions: availablePermissions,
+	}
+
+	// Search for an identical existing prompt, merge if found
+	for _, prompt := range userEntry.ByID {
+		if prompt.Snap == snap && prompt.Interface == iface && prompt.Constraints.equals(constraints) {
+			prompt.listenerReqs = append(prompt.listenerReqs, listenerReq)
```

Comment:
since we're keeping those in memory, it'd be nice to have some sort of an upper limit on how many requests can be kept

Reply:
Good point. A few considerations:

- Listener requests sit waiting in their own goroutine for a reply, so it's rather important that we don't keep too many open.
- Do we want to auto-deny new listener requests rather than adding them to this list, or replace and deny old ones (I think I favour the latter here)?
- Do we want to apply the same to prompts themselves, not just listener requests associated with a given prompt?
  - Do we want to preserve the oldest prompts and auto-deny new ones, or auto-deny old ones and preserve new ones (I think I favour the former here)?

What do you think of these @bboozzoo and @pedronis ?

Reply:
is there some sort of limiting on the kernel side or are we at the mercy of the sandboxed process? AFAIU in a stable scenario there should be only a handful of prompts and requests, shouldn't there? Unless the sandboxed process is multi threaded or uses async IO it should be blocked inside a syscall awaiting a reply to a prompt, so I think I'd start with some modest limits and then see how it works in practice. Say, 10 requests and 100 prompts?

Reply:
Just for the record, perhaps we can address this in a followup PR?

Reply:
There is no limit on the kernel side, a process could theoretically trigger as many syscalls as it can and each could result in an apparmor request.

If the current approach of starting a new goroutine to wait for a reply for each request is too heavy, we could instead actually store the received message in memory and use it later to construct and send a reply. This would also be of great benefit if we want to persist requests (and prompts) across snapd restarts. However, persisting these across snapd restarts comes with a load of complexities on both the kernel and snapd side, discussed a bit below. So if the overhead is not much different, a waiting goroutine seems much safer and simpler.

--------

When the listener is created, it opens a file descriptor on the notify.socket and registers it to receive notifications in userspace with some filter (currently empty). Associated with that fd, the kernel creates a notification queue with the given filter. When a new event triggers a notification, the kernel goes through the queues and places the notification into the first queue whose filter matches the notification.

When a listener is closed, it closes the associated fd, and the kernel redistributes any un-read notifications from that queue to the first other queue which matches the filter for the given notification.

Once a message is read from the queue by its associated listener, the kernel will not redistribute it even if that listener closes its fd before it has replied. So it would require lots of work on the kernel side to distribute notifications to multiple queues (which introduces complexity and reply races) or reclaim un-replied-to notifications which have been read by a listener which has since closed its fd without replying (not sure if this is even possible).

Reply:
I think the questions of limits to the number of outstanding prompts (and/or listeners) is a UX question as well, so I'll wait to address it until we've discussed with Desktop and Design people about this point. Persisting across restarts is another related topic we'll discuss.

Reply:
I thing given that we do a linear scan we should cap at something that is unreasonable to keep scanning this way. Lower thousands perhaps?

Reply:
I'll set the outstanding prompts per user limit to 1000, and not set a limit on number of listener requests per prompt. I think we want to deal with that in the listener itself, since optimizations there could make storing listener requests extremely lightweight. And users are unlikely to be able to answer >1000 prompts, but snapd can easily responds to a million listener requests automatically when their associated prompt is satisfied.